src/ripple/app/ledger/impl/LedgerMaster.cpp

```diff
@@ -51,6 +51,85 @@
 
 namespace ripple {
 
+namespace {
+
+//==============================================================================
+/**
+    Automatically unlocks and re-locks a unique_lock object.
+
+    This is the reverse of a std::unique_lock object - instead of locking the mutex
+    for the lifetime of this object, it unlocks it.
+
+    Make sure you don't try to unlock mutexes that aren't actually locked!
+
+    This is essentially a less-versatile boost::reverse_lock.
+
+    e.g. @code
+
+    std::mutex mut;
+
+    for (;;)
+    {
+        std::unique_lock myScopedLock{mut};
+        // mut is now locked
+
+        ... do some stuff with it locked ..
+
+        while (xyz)
+        {
+            ... do some stuff with it locked ..
+
+            ScopedUnlock unlocker{myScopedLock};
+
+            // mut is now unlocked for the remainder of this block,
+            // and re-locked at the end.
+
+            ...do some stuff with it unlocked ...
+        }  // mut gets locked here.
+
+    }  // mut gets unlocked here
+    @endcode
+*/
+template <class MutexType>
+class ScopedUnlock
+{
+    std::unique_lock<MutexType>& lock_;
+public:
+    /** Creates a ScopedUnlock.
+
+        As soon as it is created, this will unlock the unique_lock, and
+        when the ScopedLock object is deleted, the unique_lock will
+        be re-locked.
+
+        Make sure this object is created and deleted by the same thread,
+        otherwise there are no guarantees what will happen! Best just to use it
+        as a local stack object, rather than creating on the heap.
+    */
+    explicit ScopedUnlock (std::unique_lock<MutexType>& lock)
+        : lock_ (lock)
+    {
+        assert(lock_.owns_lock());
+        lock_.unlock();
+    }
+
+    ScopedUnlock (ScopedUnlock const&) = delete;
+    ScopedUnlock& operator= (ScopedUnlock const&) = delete;
+
+    /** Destructor.
+
+        The unique_lock will be locked after the destructor is called.
+
+        Make sure this object is created and deleted by the same thread,
+        otherwise there are no guarantees what will happen!
+    */
+    ~ScopedUnlock() noexcept(false)
+    {
+        lock_.lock();
+    }
+};
+
+}
+
 using namespace std::chrono_literals;
```

Comment:
_\<Looking sheepish\>_ Not on your watch, but I just noticed this `using namespace` at file scope.  That's usually considered okay in a .cpp file.  I have a minor aversion to them in _our_ .cpp files because we do unity builds.  So a `using namespace` at file scope in one .cpp file can infect a different .cpp file.

At any rate, if you wanted to get rid of this `using namespace` I'd be in favor of that.  But don't feel like that's a requirement at all.  If you're curious, here's a commit you could cherry-pick to make the appropriate changes: https://github.com/scottschurr/rippled/commit/735b9c5af8a8b0711d2b16f1b340b5e9c3f62887.

Again, this is totally at your discretion.  I'm approving your pull request regardless.

Reply:
thanks for that extra cleanup - I have folded your changes in